File: src/parsers/fb2Parser.ts
/**
 * FictionBook2 (.fb2) parser (PRD §3.1).
 *
 * One ParsedSection per top-level <section> in the source. Section title
 * comes from the FB2 <title>; "Untitled" if absent. Inline images are
 * preserved in the section HTML; the first one becomes the book cover.
 */

import JSZip from 'jszip'
import type { ParsedBook, ParsedSection, ParsedCover, ParsedImage } from './types'

const FB2_NS = 'http://www.gribuser.ru/xml/fictionbook/2.0'
const XLINK_NS = 'http://www.w3.org/1999/xlink'
import { normalizeWhitespace } from './textUtils'

const EXT_BY_MIME: Record<string, string> = {
  'image/jpeg': '.jpg',
  'image/png': '.png',
  'image/gif': '.gif',
  'image/webp': '.webp',
  'image/svg+xml': '.svg',
}

interface BinaryEntry {
  blob: Blob
  mimeType: string
  /** Stable basename used as both the OPFS filename and the `opfs:{name}` marker. */
  name: string
}

function base64ToBlob(b64: string, mimeType: string): Blob {
  const binary = atob(b64.replace(/\s/g, ''))
  const bytes = new Uint8Array(binary.length)
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i)
  return new Blob([bytes], { type: mimeType })
}

function extractBinaries(root: Document): { byHref: Map<string, BinaryEntry>; images: ParsedImage[] } {
  const byHref = new Map<string, BinaryEntry>()
  const images: ParsedImage[] = []
  const binaryEls = root.getElementsByTagNameNS(FB2_NS, 'binary')
  for (const el of Array.from(binaryEls)) {
    const id = el.getAttribute('id') ?? ''
    const contentType = el.getAttribute('content-type') ?? 'image/jpeg'
    const dataB64 = el.textContent
    if (!id || !dataB64) continue
    const blob = base64ToBlob(dataB64, contentType)
    if (!blob.size) continue
    // Use the FB2 binary id as the OPFS basename, with a sensible extension.
    const ext = EXT_BY_MIME[contentType] ?? '.bin'
    const safeId = id.replace(/[^a-zA-Z0-9._-]/g, '_')
    const name = safeId.includes('.') ? safeId : `${safeId}${ext}`
    const entry: BinaryEntry = { blob, mimeType: contentType, name }
    byHref.set(id, entry)
    byHref.set(`#${id}`, entry)
    images.push({ name, blob, mimeType: contentType })
  }
  return { byHref, images }
}

function getSectionTitle(section: Element): string {
  const titleEl = section.getElementsByTagNameNS(FB2_NS, 'title')[0]
  if (!titleEl) return ''
  const parts: string[] = []
  for (const p of Array.from(titleEl.getElementsByTagNameNS(FB2_NS, 'p'))) {
    const t = p.textContent?.trim()
    if (t) parts.push(t)
  }
  const joined = normalizeWhitespace(parts.join(' '))
  return joined || normalizeWhitespace(titleEl.textContent ?? '')
}

/** Recursively serialize an FB2 element into sanitized HTML. */
function elementToHtml(el: Element, binaries: Map<string, BinaryEntry>): string {
  const local = el.localName
  if (local === 'binary') return ''

  if (local === 'image') {
    const href =
      el.getAttributeNS(XLINK_NS, 'href') ??
      el.getAttribute('href') ??
      ''
    const entry = binaries.get(href)
    if (!entry) return ''
    return `<img src="opfs:${entry.name}" alt="" />`
  }

  // FB2 → HTML mapping
  let tag: string | null
  switch (local) {
    case 'p':
    case 'cite':
    case 'epigraph':
      tag = 'p'
      break
    case 'emphasis':
      tag = 'em'
      break
    case 'strong':
      tag = 'strong'
      break
    case 'subtitle':
      tag = 'h3'
      break
    case 'title':
      tag = 'h2'
      break
    case 'poem':
    case 'stanza':
      tag = 'blockquote'
      break
    case 'v':
      tag = 'p'
      break
    default:
      // Unknown — unwrap children
      tag = null
  }

  let inner = ''
  for (const node of Array.from(el.childNodes)) {
    if (node.nodeType === Node.TEXT_NODE) {
      inner += escapeHtml(node.textContent ?? '')
    } else if (node.nodeType === Node.ELEMENT_NODE) {
      inner += elementToHtml(node as Element, binaries)
    }
  }

  if (!tag) return inner
  return `<${tag}>${inner}</${tag}>`
}

function elementToText(el: Element, binaries: Map<string, BinaryEntry>): string {
  if (el.localName === 'binary') return ''
  if (el.localName === 'image') return ''
  void binaries
  const parts: string[] = []
  for (const node of Array.from(el.childNodes)) {
    if (node.nodeType === Node.TEXT_NODE) {
      parts.push(node.textContent ?? '')
    } else if (node.nodeType === Node.ELEMENT_NODE) {
      parts.push(elementToText(node as Element, binaries))
    }
  }
  return parts.join(' ')
}

function escapeHtml(s: string): string {
  return s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;')
}

export async function parseFb2(data: ArrayBuffer, filename?: string): Promise<ParsedBook> {
  let xmlData: ArrayBuffer = data
  const fn = (filename ?? '').toLowerCase()
  if (fn.endsWith('.fb2.zip') || fn.endsWith('.zip')) {
    const zip = await JSZip.loadAsync(data)
    const fb2File = Object.keys(zip.files).find((n) => n.toLowerCase().endsWith('.fb2'))
    if (!fb2File) throw new Error('No .fb2 file found inside the ZIP archive.')
    xmlData = await zip.files[fb2File].async('arraybuffer')
  }

  const xmlText = new TextDecoder('utf-8').decode(xmlData)
  const doc = new DOMParser().parseFromString(xmlText, 'application/xml')
  const parseError = doc.querySelector('parsererror')
  if (parseError) throw new Error(`Invalid FB2 XML: ${parseError.textContent}`)

  // Metadata
  let title = 'Untitled'
  let author = 'Unknown Author'
  const bookTitle = doc.getElementsByTagNameNS(FB2_NS, 'book-title')[0]
  if (bookTitle?.textContent?.trim()) title = bookTitle.textContent.trim()
  const authorEl = doc.getElementsByTagNameNS(FB2_NS, 'author')[0]
  if (authorEl) {
    const first = authorEl.getElementsByTagNameNS(FB2_NS, 'first-name')[0]
    const last = authorEl.getElementsByTagNameNS(FB2_NS, 'last-name')[0]
    const parts: string[] = []
    if (first?.textContent?.trim()) parts.push(first.textContent.trim())
    if (last?.textContent?.trim()) parts.push(last.textContent.trim())
    if (parts.length) author = parts.join(' ')
  }

  const { byHref: binaries, images: parsedImages } = extractBinaries(doc)

  // Cover: try <coverpage> first, then first available binary.
  let cover: ParsedCover | undefined
  const coverpage = doc.getElementsByTagNameNS(FB2_NS, 'coverpage')[0]
  if (coverpage) {
    const img = coverpage.getElementsByTagNameNS(FB2_NS, 'image')[0]
    if (img) {
      const href =
        img.getAttributeNS(XLINK_NS, 'href') ?? img.getAttribute('href') ?? ''
      const entry = binaries.get(href)
      if (entry) cover = { blob: entry.blob, mimeType: entry.mimeType }
    }
  }
  if (!cover && parsedImages.length) {
    cover = { blob: parsedImages[0].blob, mimeType: parsedImages[0].mimeType }
  }

  // Body sections
  const body = doc.getElementsByTagNameNS(FB2_NS, 'body')[0]
  const sections: ParsedSection[] = []
  if (body) {
    const topSections = Array.from(
      body.getElementsByTagNameNS(FB2_NS, 'section'),
    ).filter((s) => s.parentElement === body)

    if (!topSections.length) {
      const text = normalizeWhitespace(elementToText(body, binaries))
      const html = elementToHtml(body, binaries)
      sections.push({ title: 'Untitled', text, html })
    } else {
      for (const sec of topSections) {
        const sectionTitle = getSectionTitle(sec) || 'Untitled'
        const text = normalizeWhitespace(elementToText(sec, binaries))
        const html = elementToHtml(sec, binaries)
        sections.push({ title: sectionTitle, text, html })
      }
    }
  }

  return {
    title,
    author,
    contentType: 'text',
    sections,
    cover,
    parsedImages,
  }
}

File: backend/fb2_parser.py
"""FictionBook2 parser: extracts metadata and chapters from .fb2 files."""

from __future__ import annotations

import base64
import os
import re
import uuid
import zipfile

from lxml import etree

from backend.parser_types import InlineImage, ParsedBook, ParsedChapter, save_inline_image

_FB2_NS = "http://www.gribuser.ru/xml/fictionbook/2.0"
_XLINK_NS = "http://www.w3.org/1999/xlink"
_NS = {"fb": _FB2_NS, "l": _XLINK_NS}
_WHITESPACE_RE = re.compile(r"\s+")
_MIN_CHAPTER_LENGTH = 50


def _extract_binaries(tree: etree._Element, pub_uuid: str) -> dict[str, InlineImage]:
    """Extract all <binary> elements to disk and return id → InlineImage mapping."""
    binaries: dict[str, InlineImage] = {}
    counter = 0

    for binary_el in tree.findall(f"{{{_FB2_NS}}}binary"):
        bin_id = binary_el.get("id", "")
        content_type = binary_el.get("content-type", "image/jpeg")
        data_b64 = binary_el.text
        if not bin_id or not data_b64:
            continue

        try:
            data = base64.b64decode(data_b64.strip())
        except Exception:
            continue

        if len(data) < 500:
            continue  # Skip tiny decorations

        placeholder = f"{{{{IMG_{counter}}}}}"
        rel_path, width, height = save_inline_image(data, pub_uuid, counter, content_type)

        binaries[bin_id] = InlineImage(
            placeholder=placeholder,
            image_path=rel_path,
            alt="",
            width=width,
            height=height,
            mime_type=content_type,
        )
        # Also map with # prefix since FB2 uses href="#id"
        binaries[f"#{bin_id}"] = binaries[bin_id]
        counter += 1

    return binaries


def _get_text_with_images(
    element: etree._Element,
    binaries: dict[str, InlineImage],
    collected_images: list[InlineImage],
) -> str:
    """Recursively extract text from an element, replacing <image> refs with placeholders."""
    tag = element.tag
    if tag == f"{{{_FB2_NS}}}binary":
        return ""

    # Handle <image> elements — replace with placeholder
    if tag == f"{{{_FB2_NS}}}image":
        href = element.get(f"{{{_XLINK_NS}}}href", "") or element.get("href", "")
        if href and href in binaries:
            img = binaries[href]
            if img not in collected_images:
                collected_images.append(img)
            return f" {img.placeholder} "
        return ""

    parts: list[str] = []
    if element.text:
        parts.append(element.text)
    for child in element:
        parts.append(_get_text_with_images(child, binaries, collected_images))
        if child.tail:
            parts.append(child.tail)
    return " ".join(parts)


def _extract_section_title(section: etree._Element) -> str:
    """Get the title from a <section><title>... element."""
    title_el = section.find("fb:title", _NS)
    if title_el is not None:
        # Extract title text without image placeholders
        parts: list[str] = []
        for p in title_el.findall(f".//{{{_FB2_NS}}}p"):
            if p.text:
                parts.append(p.text.strip())
        result = _WHITESPACE_RE.sub(" ", " ".join(parts)).strip()
        if result:
            return result
        # Fallback: all text from title element
        return _WHITESPACE_RE.sub(" ", "".join(title_el.itertext())).strip()
    return ""


def _parse_sections(
    body: etree._Element,
    binaries: dict[str, InlineImage],
) -> list[ParsedChapter]:
    """Walk <section> elements and produce chapters with inline images."""
    chapters: list[ParsedChapter] = []
    sections = body.findall("fb:section", _NS)

    if not sections:
        collected: list[InlineImage] = []
        text = _WHITESPACE_RE.sub(
            " ", _get_text_with_images(body, binaries, collected),
        ).strip()
        if len(text) >= _MIN_CHAPTER_LENGTH:
            chapters.append(ParsedChapter(
                title="Chapter 1", text=text, inline_images=collected,
            ))
        return chapters

    for i, section in enumerate(sections):
        title = _extract_section_title(section) or f"Chapter {i + 1}"

        subsections = section.findall("fb:section", _NS)
        if subsections:
            for j, sub in enumerate(subsections):
                sub_title = _extract_section_title(sub)
                collected: list[InlineImage] = []
                sub_text = _WHITESPACE_RE.sub(
                    " ", _get_text_with_images(sub, binaries, collected),
                ).strip()
                if len(sub_text) >= _MIN_CHAPTER_LENGTH:
                    full_title = f"{title} - {sub_title}" if sub_title else f"{title} ({j + 1})"
                    chapters.append(ParsedChapter(
                        title=full_title, text=sub_text, inline_images=collected,
                    ))
        else:
            collected: list[InlineImage] = []
            text = _WHITESPACE_RE.sub(
                " ", _get_text_with_images(section, binaries, collected),
            ).strip()
            if len(text) >= _MIN_CHAPTER_LENGTH:
                chapters.append(ParsedChapter(
                    title=title, text=text, inline_images=collected,
                ))

    return chapters


def parse_fb2(file_path: str) -> ParsedBook:
    """Parse a FictionBook2 (.fb2 or .fb2.zip) file into chapters with images."""
    if file_path.lower().endswith(".fb2.zip") or file_path.lower().endswith(".zip"):
        with zipfile.ZipFile(file_path, "r") as zf:
            fb2_names = [n for n in zf.namelist() if n.lower().endswith(".fb2")]
            if not fb2_names:
                raise ValueError("No .fb2 file found inside the ZIP archive.")
            xml_data = zf.read(fb2_names[0])
    else:
        with open(file_path, "rb") as f:
            xml_data = f.read()

    try:
        tree = etree.fromstring(xml_data)
    except etree.XMLSyntaxError as exc:
        raise ValueError(f"Invalid FB2 XML: {exc}") from exc

    # Metadata from <title-info>
    title_info = tree.find(".//fb:description/fb:title-info", _NS)

    title = "Untitled"
    author = "Unknown Author"

    if title_info is not None:
        book_title = title_info.find("fb:book-title", _NS)
        if book_title is not None and book_title.text:
            title = book_title.text.strip()

        author_el = title_info.find("fb:author", _NS)
        if author_el is not None:
            first = author_el.find("fb:first-name", _NS)
            last = author_el.find("fb:last-name", _NS)
            parts = []
            if first is not None and first.text:
                parts.append(first.text.strip())
            if last is not None and last.text:
                parts.append(last.text.strip())
            if parts:
                author = " ".join(parts)

    # Extract binary images
    pub_uuid = str(uuid.uuid4())
    binaries = _extract_binaries(tree, pub_uuid)

    # Parse body sections
    body = tree.find("fb:body", _NS)
    if body is None:
        return ParsedBook(title=title, author=author, chapters=[])

    chapters = _parse_sections(body, binaries)

    return ParsedBook(title=title, author=author, chapters=chapters)
